File: 32 - Guess the Number Game/Golang/main.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

func getInt(output string) int {
	fmt.Print(output)
	var temp int
	fmt.Scanf("%d", &temp)
	return temp
}

func game() {
	difficulty := getInt("What difficulty do you want to play on (1,2,3)? ")
	rand.Seed(time.Now().UnixNano())
	number := rand.Intn(int(math.Pow(10, float64(difficulty)))-1) + 1 // range 1 - 10*difficulty
	nGuesses := 0
	for {
		nGuesses += 1
		guess := getInt("What is your guess? ")
		if guess > number {
			fmt.Println("Too high.")
		} else if guess < number {
			fmt.Println("Too low.")
		} else {
			break
		}
	}

	fmt.Printf("It took you %d guesses to guess the number\n", nGuesses)
}

func main() {
    game()
    for {
        var play_again string
        fmt.Print("Want to play again? y/n")
        fmt.Scanf("%s", &play_again)
        if play_again == "y" {
            game()
        } else if play_again == "n" {
            fmt.Println("Bye!")
            break
        } else {
            fmt.Println("Not a valid input.")
        }
    }
}

File: 32 - Guess the Number Game/C++/main.cpp
#include <cmath>
#include <iostream>
#include <random>

// returns a random integer in the range passed
int random_int(int low, int high)
{
    std::random_device                 rand_dev;
    std::mt19937                       generator(rand_dev());
    std::uniform_int_distribution<int> distr(low, high);
    return distr(generator);
}

int get_int(std::string output)
{
    int temp;
    std::cout << output;
    std::cin >> temp;
    return temp;
}

void game()
{
    int n_guesses = 0;  // first guess is outside the loop
    int difficulty = get_int("What difficulty? (1,2,3): ");
    int number    = random_int(0, std::pow(10, difficulty));
    while (true) {
        n_guesses++;
        int guess = get_int("What is your guess? ");
        if (guess < number) {
            std::cout << "Too low." << std::endl;
            continue;
        } else if (guess > number) {
            std::cout << "Too high." << std::endl;
            continue;
        } else {
            break;
        }
    }
    std::cout << "It took you " << n_guesses << " guesses to get the number." << std::endl;
}

int main() {
    game();
    while (true) {
        std::string play_again;
        std::cout << "Play again? y/n ";
        std::cin >> play_again;
        if (play_again == "y") {
            game();
            continue;
        } else if (play_again == "n") {
            std::cout << "Bye!" << std::endl;
            break;
        } else {
            std::cout << "Not a valid input." << std::endl;
            continue;
        }
    }
    return 0;
}
